sktime/utils/_testing/deep_equals.py

```diff
@@ -30,52 +32,93 @@ def deep_equals(x, y):
 
     Parameters
     ----------
-    x: object
-    y: object
+    x : object
+    y : object
+    return_msg : bool, optional, default=False
+        whether to return informative message about what is not equal
 
     Returns
     -------
-    bool - True if x and y are equal in value
+    is_equal: bool - True if x and y are equal in value
         x and y do not need to be equal in reference
+    msg : str, only returned if return_msg = True
+        indication of what is the reason for not being equal
+            concatenation of the following strings:
+            .type - type is not equal
+            .len - length is not equal
+            .value - value is not equal
+            .keys - if dict, keys of dict are not equal
+                    if class/object, names of attributes and methods are not equal
+            .dtype - dtype of pandas or numpy object is not equal
+            .index - index of pandas object is not equal
+            .series_equals, .df_equals - .equals call on pandas objects returns False
+            [i] - if tuple/list: i-th element not equal
+            [key] - if dict: value at key is not equal
+            [colname] - if pandas.DataFrame: column with name colname is not equal
+            != - call to generic != returns False
     """
+
+    def ret(is_equal, msg):
```

Comment:
This is a little confusing since it takes two args and an additional one from the closure, any thoughts on refactoring this to be a bit easier to read without jumping back and forth?

Reply:
no good ideas. We could have a third argument, and move it out of the scope, but that would make the return statements hard to read. Any suggestions?